src/blog/2023/06/node-red-ha-flowforge.md

```diff
@@ -0,0 +1,99 @@
+
+This article is not an attempt to fully load-test FlowForge's performance with HA, the loads used were synthetic but should correlate strongly with real world performance of Node-RED applications.
+
+### How does the preview build of FlowForge's HA work?
+
+When HA is enabled in FlowForge for a Node-RED instance, various things change under the hood. We need to consider these changes when writing Node-RED flows for HA. The main points we need to consider are:
+
+##### HA runs multiple copies of the same instance
+
+Once HA is enabled, FlowForge creates multiple copies of your instance (2 at the time of writing). These copies all run in parallel. Any traffic to your instance will be load-balanced so that the traffic is shared between all copies.
+
+The load-balanced approach to HA creates a challenge when designing your applications, because it means you cannot:
+
+* determine which copy of your flows is dealing with a given message.
+* configure the copies of the flows to communicate to each other directly.
+* determine how many copies are running at any given moment in time.
+
+These factors need to be considered when designing your applications. They are not negative as such, but require a different approach when creating an application which will run in a single Node-RED instance.
+
+##### Limited access to the Node-RED editor interface
+
+Once HA is enabled, you cannot edit the flows in a Node-RED instance directly, you can't even access the interface for Node-RED. This is a deliberate limitation, if you were to edit the flows directly, you would end up with different versions of your instance running at the same time.
+
+Don't worry, we have a solution. To update your instance you can use our [DevOps pipeline feature](https://flowforge.com/docs/user/devops-pipelines/#devops-pipelines) to promote the flows on a Node-RED instance to your HA instance. Editing your flows in this way allows FlowForge to ensure that the same flows are running on all your Node-RED HA copies.
+
+### Let's build something
+
+To explore FlowForge's HA feature preview, I am going to build a basic load testing application. The application will make an API call to a Node-RED instance which will in turn reply with a small payload. 
+
+I'm not looking to methodically load-test HA on FlowForge Cloud here, but I want to make sure I can see the impact of HA on the application. To that end, I am going to create a synthetic bottleneck on the flows running in HA. In simple terms, I will use a delay node which only allows one message to pass through every 100 milliseconds. I'm going to queue up any messages that arrive in the mean-time. This is a reasonable simulation of what we might see if each connection to the HA instance was in turn completing an interaction with a database or MQTT broker.
```

Comment:
I don't understand how introducing a synthetic bottleneck simulates HA of flows or the comment about the interaction with a database or MQTT broker? HiveMQ recently published a benchmark of showing 200 million simultaneous connections. I don't believe they introduced a bottleneck for this simulation?

For me HA is defined by:
- a large number of different connections trying to connect at the same time
- and/or smaller number of connections trying to sending many messages at the same time.